File: yeet-synth/synth/evaluator.py
"""
Pass/fail evaluation model.

At the end of a run, the Evaluator checks all collected metrics and mesh
results against configured thresholds and produces a formal verdict.

Exit codes:
  0  — all checks passed
  1  — one or more FAIL checks
  2  — all checks passed but WARNs present
  3  — evaluation could not run (insufficient data)
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

from .config import ThresholdConfig
from .mesh import MeshCheckResult, CheckStatus
from .metrics import MetricsCollector


class Verdict(Enum):
    PASS = "PASS"
    WARN = "WARN"
    FAIL = "FAIL"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"


@dataclass
class CheckResult:
    name: str
    verdict: Verdict
    message: str
    observed: Optional[float] = None
    threshold: Optional[float] = None
    unit: str = ""


@dataclass
class EvaluationResult:
    verdict: Verdict
    checks: list[CheckResult] = field(default_factory=list)
    confidence: float = 0.0       # 0-100: how many requests were sampled
    exit_code: int = 0

    def add(self, check: CheckResult) -> None:
        self.checks.append(check)

    def fails(self) -> list[CheckResult]:
        return [c for c in self.checks if c.verdict == Verdict.FAIL]

    def warns(self) -> list[CheckResult]:
        return [c for c in self.checks if c.verdict == Verdict.WARN]

    def as_dict(self) -> dict:
        return {
            "verdict": self.verdict.value,
            "exit_code": self.exit_code,
            "confidence": self.confidence,
            "checks": [
                {
                    "name": c.name,
                    "verdict": c.verdict.value,
                    "message": c.message,
                    "observed": c.observed,
                    "threshold": c.threshold,
                    "unit": c.unit,
                }
                for c in self.checks
            ],
        }


_MIN_REQUESTS_FOR_CONFIDENCE = 50


class Evaluator:

    def __init__(self, thresholds: ThresholdConfig) -> None:
        self._t = thresholds

    def evaluate(
        self,
        metrics: MetricsCollector,
        mesh_results: Optional[list[MeshCheckResult]] = None,
        chaos_results: Optional[list] = None,
    ) -> EvaluationResult:
        snap = metrics.snapshot()
        total = sum(m.total for m in snap.values())

        if total < 10:
            return EvaluationResult(
                verdict=Verdict.INSUFFICIENT_DATA,
                confidence=0.0,
                exit_code=3,
            )

        # Confidence: saturates at 1000 requests
        confidence = min(100.0, total / 10.0)
        result = EvaluationResult(verdict=Verdict.PASS, confidence=confidence)

        # ── Global error rate ────────────────────────────────────────────────
        error_rate = metrics.global_error_rate()
        result.add(CheckResult(
            name="global_error_rate",
            verdict=Verdict.FAIL if error_rate > self._t.max_error_rate else Verdict.PASS,
            message=f"Error rate {error_rate:.2%} vs threshold {self._t.max_error_rate:.2%}",
            observed=error_rate * 100,
            threshold=self._t.max_error_rate * 100,
            unit="%",
        ))

        # ── Global p99 latency ───────────────────────────────────────────────
        p99 = metrics.global_p99()
        result.add(CheckResult(
            name="global_p99_latency",
            verdict=Verdict.FAIL if p99 > self._t.p99_latency_ms else
                    Verdict.WARN if p99 > self._t.p95_latency_ms else Verdict.PASS,
            message=f"p99 latency {p99:.0f}ms vs threshold {self._t.p99_latency_ms:.0f}ms",
            observed=p99,
            threshold=self._t.p99_latency_ms,
            unit="ms",
        ))

        # ── Per-endpoint checks ──────────────────────────────────────────────
        for ep_name, em in snap.items():
            if em.total < 5:
                continue

            # Timeout rate
            if em.timeout_rate > self._t.max_timeout_rate:
                result.add(CheckResult(
                    name=f"timeout_rate:{ep_name}",
                    verdict=Verdict.WARN,
                    message=f"{ep_name}: timeout rate {em.timeout_rate:.2%} > threshold",
                    observed=em.timeout_rate * 100,
                    threshold=self._t.max_timeout_rate * 100,
                    unit="%",
                ))

            # Auth failure rate
            if em.total > 0 and em.auth_failures / em.total > self._t.max_auth_failure_rate:
                result.add(CheckResult(
                    name=f"auth_failure_rate:{ep_name}",
                    verdict=Verdict.WARN,
                    message=f"{ep_name}: auth failure rate {em.auth_failures / em.total:.2%}",
                    observed=em.auth_failures / em.total * 100,
                    threshold=self._t.max_auth_failure_rate * 100,
                    unit="%",
                ))

            # Retry amplification
            if em.avg_attempt_count > self._t.retry_amplification_threshold:
                result.add(CheckResult(
                    name=f"retry_amplification:{ep_name}",
                    verdict=Verdict.WARN,
                    message=(
                        f"{ep_name}: avg attempt count {em.avg_attempt_count:.2f} "
                        f"> threshold {self._t.retry_amplification_threshold:.1f} "
                        "(possible retry storm)"
                    ),
                    observed=em.avg_attempt_count,
                    threshold=self._t.retry_amplification_threshold,
                ))

            # Trace propagation
            if em.trace_sent > 10 and em.trace_propagation_rate < self._t.min_trace_propagation_rate:
                result.add(CheckResult(
                    name=f"trace_propagation:{ep_name}",
                    verdict=Verdict.WARN,
                    message=(
                        f"{ep_name}: trace propagation {em.trace_propagation_rate:.0%} "
                        f"< threshold {self._t.min_trace_propagation_rate:.0%}"
                    ),
                    observed=em.trace_propagation_rate * 100,
                    threshold=self._t.min_trace_propagation_rate * 100,
                    unit="%",
                ))

        # ── Mesh checks ──────────────────────────────────────────────────────
        if mesh_results:
            for mr in mesh_results:
                if mr.status == CheckStatus.SKIP:
                    continue
                result.add(CheckResult(
                    name=f"mesh:{mr.name}",
                    verdict=Verdict.FAIL if mr.status == CheckStatus.FAIL else
                            Verdict.WARN if mr.status == CheckStatus.WARN else Verdict.PASS,
                    message=mr.message,
                ))

        # ── Chaos checks ─────────────────────────────────────────────────────
        if chaos_results:
            for cr in chaos_results:
                result.add(CheckResult(
                    name=f"chaos:{cr.scenario}",
                    verdict=Verdict.PASS if cr.passed else Verdict.FAIL,
                    message=cr.note or f"Expected {cr.expected_status}, got {cr.status_code}",
                    observed=float(cr.status_code),
                    threshold=float(cr.expected_status),
                ))

        # ── Final verdict ────────────────────────────────────────────────────
        if any(c.verdict == Verdict.FAIL for c in result.checks):
            result.verdict = Verdict.FAIL
            result.exit_code = 1
        elif any(c.verdict == Verdict.WARN for c in result.checks):
            result.verdict = Verdict.WARN
            result.exit_code = 2
        else:
            result.verdict = Verdict.PASS
            result.exit_code = 0

        return result

File: y_eet-synth/internal/evaluator/evaluator.go
// Package evaluator implements pass/fail evaluation of collected metrics.
package evaluator

import (
	"fmt"

	"y_eet-synth/internal/config"
	"y_eet-synth/internal/metrics"
)

// Verdict is the overall evaluation outcome.
type Verdict string

const (
	VerdictPass             Verdict = "PASS"
	VerdictWarn             Verdict = "WARN"
	VerdictFail             Verdict = "FAIL"
	VerdictInsufficientData Verdict = "INSUFFICIENT_DATA"
)

// CheckVerdict is the result of a single threshold check.
type CheckVerdict string

const (
	CheckPass CheckVerdict = "PASS"
	CheckWarn CheckVerdict = "WARN"
	CheckFail CheckVerdict = "FAIL"
)

// Check is the result of one threshold check.
type Check struct {
	Name      string       `json:"name"`
	Verdict   CheckVerdict `json:"verdict"`
	Message   string       `json:"message"`
	Observed  float64      `json:"observed,omitempty"`
	Threshold float64      `json:"threshold,omitempty"`
	Unit      string       `json:"unit,omitempty"`
}

// MeshCheckStatus represents a mesh validation check outcome.
type MeshCheckStatus string

const (
	MeshPass MeshCheckStatus = "PASS"
	MeshWarn MeshCheckStatus = "WARN"
	MeshFail MeshCheckStatus = "FAIL"
	MeshSkip MeshCheckStatus = "SKIP"
)

// MeshResult is the result of one mesh validation check.
type MeshResult struct {
	Name    string          `json:"name"`
	Status  MeshCheckStatus `json:"status"`
	Message string          `json:"message"`
	Details map[string]interface{} `json:"details,omitempty"`
}

// ChaosResult is the result of one chaos scenario.
type ChaosResult struct {
	Scenario       string `json:"scenario"`
	Passed         bool   `json:"passed"`
	ExpectedStatus int    `json:"expected_status"`
	StatusCode     int    `json:"status_code"`
	Note           string `json:"note,omitempty"`
}

// Result is the final evaluation output.
type Result struct {
	Verdict    Verdict  `json:"verdict"`
	ExitCode   int      `json:"exit_code"`
	Confidence float64  `json:"confidence"`
	Checks     []Check  `json:"checks"`
}

func (r *Result) Fails() []Check {
	var out []Check
	for _, c := range r.Checks {
		if c.Verdict == CheckFail {
			out = append(out, c)
		}
	}
	return out
}

func (r *Result) Warns() []Check {
	var out []Check
	for _, c := range r.Checks {
		if c.Verdict == CheckWarn {
			out = append(out, c)
		}
	}
	return out
}

// Evaluator runs threshold checks against collected metrics.
type Evaluator struct {
	t config.ThresholdConfig
}

// New creates an Evaluator.
func New(t config.ThresholdConfig) *Evaluator {
	return &Evaluator{t: t}
}

// Evaluate produces a Result from the collected metrics and optional mesh/chaos outcomes.
func (e *Evaluator) Evaluate(
	m *metrics.Collector,
	meshResults []MeshResult,
	chaosResults []ChaosResult,
) Result {
	snap := m.Snapshot()
	total := 0
	for _, em := range snap {
		total += em.Total
	}

	if total < 10 {
		return Result{
			Verdict:    VerdictInsufficientData,
			ExitCode:   3,
			Confidence: 0,
		}
	}

	confidence := float64(total) / 10.0
	if confidence > 100 {
		confidence = 100
	}

	result := Result{
		Verdict:    VerdictPass,
		Confidence: confidence,
	}

	// Global error rate
	errRate := m.GlobalErrorRate()
	result.Checks = append(result.Checks, Check{
		Name: "global_error_rate",
		Verdict: check(errRate > e.t.MaxErrorRate, errRate > e.t.MaxErrorRate/2),
		Message: fmt.Sprintf("error rate %.2f%% vs threshold %.2f%%",
			errRate*100, e.t.MaxErrorRate*100),
		Observed:  errRate * 100,
		Threshold: e.t.MaxErrorRate * 100,
		Unit:      "%",
	})

	// Global p99 latency
	p99 := m.GlobalP99()
	result.Checks = append(result.Checks, Check{
		Name: "global_p99_latency",
		Verdict: func() CheckVerdict {
			if p99 > e.t.P99LatencyMs {
				return CheckFail
			}
			if p99 > e.t.P95LatencyMs {
				return CheckWarn
			}
			return CheckPass
		}(),
		Message: fmt.Sprintf("p99 %.0fms vs threshold %.0fms", p99, e.t.P99LatencyMs),
		Observed:  p99,
		Threshold: e.t.P99LatencyMs,
		Unit:      "ms",
	})

	// Per-endpoint checks
	for epName, em := range snap {
		if em.Total < 5 {
			continue
		}

		// Timeout rate
		if em.TimeoutRate() > e.t.MaxTimeoutRate {
			result.Checks = append(result.Checks, Check{
				Name:    "timeout_rate:" + epName,
				Verdict: CheckWarn,
				Message: fmt.Sprintf("%s: timeout rate %.2f%% > threshold", epName,
					em.TimeoutRate()*100),
				Observed:  em.TimeoutRate() * 100,
				Threshold: e.t.MaxTimeoutRate * 100,
				Unit:      "%",
			})
		}

		// Auth failure rate
		if em.Total > 0 {
			authRate := float64(em.AuthFailures) / float64(em.Total)
			if authRate > e.t.MaxAuthFailureRate {
				result.Checks = append(result.Checks, Check{
					Name:    "auth_failure_rate:" + epName,
					Verdict: CheckWarn,
					Message: fmt.Sprintf("%s: auth failure rate %.2f%%", epName, authRate*100),
					Observed:  authRate * 100,
					Threshold: e.t.MaxAuthFailureRate * 100,
					Unit:      "%",
				})
			}
		}

		// Retry amplification
		if em.AvgAttemptCount() > e.t.RetryAmplificationThreshold {
			result.Checks = append(result.Checks, Check{
				Name:    "retry_amplification:" + epName,
				Verdict: CheckWarn,
				Message: fmt.Sprintf("%s: avg attempt count %.2f > threshold %.1f",
					epName, em.AvgAttemptCount(), e.t.RetryAmplificationThreshold),
				Observed:  em.AvgAttemptCount(),
				Threshold: e.t.RetryAmplificationThreshold,
			})
		}

		// Trace propagation
		if em.TraceSent > 10 && em.TracePropagationRate() < e.t.MinTracePropagationRate {
			result.Checks = append(result.Checks, Check{
				Name:    "trace_propagation:" + epName,
				Verdict: CheckWarn,
				Message: fmt.Sprintf("%s: trace propagation %.0f%% < threshold %.0f%%",
					epName, em.TracePropagationRate()*100, e.t.MinTracePropagationRate*100),
				Observed:  em.TracePropagationRate() * 100,
				Threshold: e.t.MinTracePropagationRate * 100,
				Unit:      "%",
			})
		}
	}

	// Mesh results
	for _, mr := range meshResults {
		if mr.Status == MeshSkip {
			continue
		}
		v := CheckPass
		if mr.Status == MeshFail {
			v = CheckFail
		} else if mr.Status == MeshWarn {
			v = CheckWarn
		}
		result.Checks = append(result.Checks, Check{
			Name:    "mesh:" + mr.Name,
			Verdict: v,
			Message: mr.Message,
		})
	}

	// Chaos results
	for _, cr := range chaosResults {
		v := CheckPass
		if !cr.Passed {
			v = CheckFail
		}
		msg := cr.Note
		if msg == "" {
			msg = fmt.Sprintf("expected %d, got %d", cr.ExpectedStatus, cr.StatusCode)
		}
		result.Checks = append(result.Checks, Check{
			Name:      "chaos:" + cr.Scenario,
			Verdict:   v,
			Message:   msg,
			Observed:  float64(cr.StatusCode),
			Threshold: float64(cr.ExpectedStatus),
		})
	}

	// Final verdict
	hasFail := false
	hasWarn := false
	for _, c := range result.Checks {
		if c.Verdict == CheckFail {
			hasFail = true
		}
		if c.Verdict == CheckWarn {
			hasWarn = true
		}
	}
	switch {
	case hasFail:
		result.Verdict = VerdictFail
		result.ExitCode = 1
	case hasWarn:
		result.Verdict = VerdictWarn
		result.ExitCode = 2
	default:
		result.Verdict = VerdictPass
		result.ExitCode = 0
	}

	return result
}

func check(fail, _ bool) CheckVerdict {
	if fail {
		return CheckFail
	}
	return CheckPass
}
